pkg/runtime/controller.go

```diff
@@ -0,0 +1,424 @@
+	instanceConfig := make(map[string]*cpb.Instance)
+
+	// first pass get all the validated instance and handler references
+	for k, cfg := range c.configState {
+		if _, found := c.templateInfo[k.Kind]; !found {
+			continue
+		}
+		instanceConfig[k.String()] = &cpb.Instance{
+			Name:     k.Name,
+			Template: k.Kind,
+			Params:   cfg,
+		}
+	}
+	return instanceConfig
+}
+
+// validHandlerConfigs returns handlerConfigs from the configState that
+// point to valid adapters.
+func (c *Controller) validHandlerConfigs() map[string]*cpb.Handler {
+	handlerConfig := make(map[string]*cpb.Handler)
+	for k, cfg := range c.configState {
+		if _, found := c.adapterInfo[k.Kind]; !found {
+			continue
+		}
+		handlerConfig[k.String()] = &cpb.Handler{
+			Name:    k.Name,
+			Adapter: k.Kind,
+			Params:  cfg,
+		}
+	}
+	return handlerConfig
+}
+
+type rulesByName map[string]*Rule
+
+// rulesByName indexed by namespace
+type rulesMapByNamespace map[string]rulesByName
+
+// rulesListByNamespace is the type needed by resolver.
+type rulesListByNamespace map[string][]*Rule
+
+// convertToRuntimeRules converts internal rules to the format that resolver needs.
+func convertToRuntimeRules(ruleConfig rulesMapByNamespace) (rulesListByNamespace, int) {
+	// convert rules
+	nrules := 0
+	rules := make(rulesListByNamespace)
+	for ns, nsmap := range ruleConfig {
+		rulesArr := make([]*Rule, 0, len(nsmap))
+		for _, rule := range nsmap {
+			rulesArr = append(rulesArr, rule)
+		}
+		rules[ns] = rulesArr
+		nrules += len(rulesArr)
+	}
+	return rules, nrules
+}
+
+// processRules builds the current consistent view of the rules keyed by Namespace and then Name.
+// ht (handlerTable) keeps track of handler-instance association.
+func (c *Controller) processRules(handlerConfig map[string]*cpb.Handler,
+	instanceConfig map[string]*cpb.Instance, ht *handlerTable) rulesMapByNamespace {
+	// current consistent view of the rules
+	// keyed by Namespace and then Name.
+	ruleConfig := make(rulesMapByNamespace)
+
+	// check rules and ensure only good handlers and instances are used.
+	// record handler - instance associations
+	for k, cfg := range c.configState {
+		if k.Kind != RulesKind {
+			continue
+		}
+		rulec := cfg.(*cpb.Rule)
+		rule := &Rule{
+			selector: rulec.Selector,
+			name:     k.Name,
+		}
+		acts := c.processActions(rulec.Actions, handlerConfig, instanceConfig, ht)
+
+		rn := ruleConfig[k.Namespace]
+		if rn == nil {
+			rn = make(map[string]*Rule)
+			ruleConfig[k.Namespace] = rn
+		}
+		ruleActions := make(map[adptTmpl.TemplateVariety][]*Action)
+		for vr, amap := range acts {
+			for _, cf := range amap {
+				ruleActions[vr] = append(ruleActions[vr], cf)
+			}
+		}
+		rule.actions = ruleActions
+		rn[k.Name] = rule
+	}
+
+	return ruleConfig
+}
+
+var cleanupSleepTime = 500 * time.Millisecond
+
+// processActions prunes actions that lack referential integrity and associate instances with
+// handlers that are later used to create new handlers.
```

Comment:
Consider a comment..: The reference to the built handler objects is not done in this method. it is done later in combineActionsToHandler done after handlers are built.